sdk/src/main/scala/com/horizen/account/state/StateTransition.scala

```diff
@@ -105,10 +108,20 @@ class StateTransition(view: AccountStateView, messageProcessors: Seq[MessageProc
   private def refundGas(msg: Message, gas: GasPool): Unit = {
     val usedGas = msg.getGasLimit.subtract(gas.getGas)
     // cap gas refund to a quotient of the used gas
-    gas.addGas(view.getRefund.min(usedGas.divide(GasUtil.RefundQuotientEIP3529)))
+    val refGas = view.getRefund.min(usedGas.divide(GasUtil.RefundQuotientEIP3529))
+    gas.addGas(refGas)
     // return funds for remaining gas, exchanged at the original rate.
     val remaining = gas.getGas.multiply(msg.getGasPrice)
-    view.addBalance(msg.getFrom.address(), remaining)
+    val sender = msg.getFrom.address()
+    log.debug(s"gas used: $usedGas, remaining gas: ${gas.getGas}, refGas: $refGas, gasPrice: ${msg.getGasPrice}")
+    log.debug(s"now refunding $remaining to sender ${BytesUtils.toHexString(sender)}")
+    if (remaining.compareTo(BigInteger.ZERO) > 0) {
+      view.addBalance(sender, remaining)
+      // the addBalance op subtracts gas, we have to restore it here otherwise remaining gas balance is wrong
```

Comment:
Good point! The gas consumption should be disabled before we even call `refundGas`. Except for the intrinsic gas, gas should only be consumed from within the message processors. Here we already arrived at a point where "out of gas" should not happen anymore. Currently we make sure gas tracking is disabled in `applyTransaction`, lets move it to the finally block in `transition` before we call `refundGas`. What do you think?

Reply:
Agree, the modification you suggest is much cleaner. Done.